frame/support/procedural/src/benchmark.rs

```diff
@@ -0,0 +1,786 @@
+			params,
+			setup_stmts: Vec::from(&item_fn.block.stmts[0..i]),
+			call_def,
+			verify_stmts: Vec::from(&item_fn.block.stmts[(i + 1)..item_fn.block.stmts.len()]),
+			extra,
+			skip_meta,
+		})
+	}
+}
+
+/// Parses and expands a `#[benchmarks]` or `#[instance_benchmarks]` invocation
+pub fn benchmarks(attrs: TokenStream, tokens: TokenStream, instance: bool) -> TokenStream {
+	let module = parse_macro_input!(tokens as ItemMod);
+	let mod_span = module.span();
+	let where_clause = match syn::parse::<Nothing>(attrs.clone()) {
+		Ok(_) => quote!(),
+		Err(_) => parse_macro_input!(attrs as WhereClause).predicates.to_token_stream(),
+	};
+	let mod_vis = module.vis;
+	let mod_name = module.ident;
+	let mut expanded_stmts: Vec<TokenStream2> = Vec::new();
+	let mut benchmark_defs: Vec<BenchmarkDef> = Vec::new();
+	let mut benchmark_names: Vec<Ident> = Vec::new();
+	let mut extra_benchmark_names: Vec<Ident> = Vec::new();
+	let mut skip_meta_benchmark_names: Vec<Ident> = Vec::new();
+	let Some(mut content) = module.content else {
+		// this will compile error already because attributes attached to empty modules are unstable
+		// but including error anyway to make this future-proof
+		return quote_spanned!(mod_span=> "Module cannot be empty!".to_compile_error()).into()
+	};
+	for stmt in &mut content.1 {
+		let mut push_stmt = || {
+			expanded_stmts.push(stmt.to_token_stream());
+		};
+		let Item::Fn(mut func) = stmt.clone() else { push_stmt(); continue; };
+		for (i, attr) in (&func.attrs.clone()).iter().enumerate() {
+			let Some(seg) = attr.path.segments.last() else { push_stmt(); continue; };
+			let Ok(_) = syn::parse::<keywords::benchmark>(seg.ident.to_token_stream().into()) else { push_stmt(); continue; };
+			let tokens = attr.tokens.to_token_stream().into();
+			let args = parse_macro_input!(tokens as BenchmarkAttrs);
+
+			// consume #[benchmark] attr
+			func.attrs.remove(i);
+
+			// parse benchmark def
+			let benchmark_def = match BenchmarkDef::from(&func, args.extra, args.skip_meta) {
+				Ok(def) => def,
+				Err(err) => return err.to_compile_error().into(),
+			};
+
+			// expand benchmark
+			let expanded = expand_benchmark(
+				benchmark_def.clone(),
+				&func.sig.ident,
+				instance,
+				where_clause.clone(),
+			);
+
+			// record benchmark name
+			let name = func.sig.ident;
+
+			// process name vecs
+			benchmark_names.push(name.clone());
+			if benchmark_def.extra {
+				extra_benchmark_names.push(name.clone());
+			}
+			if benchmark_def.skip_meta {
+				skip_meta_benchmark_names.push(name.clone())
+			}
+
+			expanded_stmts.push(expanded);
+			benchmark_defs.push(benchmark_def);
+			break
+		}
+	}
+
+	// generics
+	let generics = match instance {
+		false => quote!(T),
+		true => quote!(T, I),
+	};
+	let full_generics = match instance {
+		false => quote!(T: Config),
+		true => quote!(T: Config<I>, I: 'static),
+	};
+
+	let krate = quote!(::frame_benchmarking);
+	let support = quote!(#krate::frame_support);
```

Comment:
We can't just assume that these crate names are `frame_benchmarking` and `frame_support`, because they may have been renamed in the `Cargo.toml` file. There is a utility function that allows us to look for the actual crate name; look for `generate_crate_access_2018`.